lib/actions/ExtensionAction.js

```diff
@@ -475,6 +518,223 @@ class ExtensionAction {
+   * @param {Boolean} [options.force] Forces creation of version without asking user
+   * @returns {Promise}
+   */
+  async uploadExtension (args = {}, options = {}) {
+    await this.userSettings.validate()
+    await this.appSettings.validate()
+
+    const isTheme = args.isTheme === true
+    const extensionsDirectory = path.join(this.appSettings.getApplicationFolder(), EXTENSIONS_FOLDER)
+
+    const extensionPropeties = await this._getAllExtensionProperties()
+    if (extensionPropeties.length === 0) {
+      return logger.warn(`There are no extensions in folder './${path.relative(process.cwd(), extensionsDirectory)}'`)
+    }
+
+    let extensionId
+    let extensionVersion
+    let extensionDir
+    let extensionVersionExists
+
+    if (args.extension) {
+      const extension = extensionPropeties.find(({ dir }) => dir === args.extension)
+
+      if (!extension) {
+        throw new Error(`${isTheme ? 'Theme' : 'Extension'} folder ${args.extension} does not exist`)
+      }
+
+      if (isTheme && extension.isTheme !== true) {
+        logger.warn(`Extension ${extension.id} exists but is not marked as theme (type: "theme")`)
+      }
+
+      extensionId = extension.id
+      extensionDir = extension.dir
+    } else {
+      const choices = extensionPropeties
+        .filter((extension) => !isTheme || extension.isTheme)
+        .map(({ id }) => ({ name: id }))
+
+      if (choices.length === 0) {
+        return logger.warn(`There are no themes in folder './${path.relative(process.cwd(), extensionsDirectory)}'`)
+      }
+
+      const selectedId = (await inquirer.prompt({
+        type: 'list',
+        name: 'extensionId',
+        message: `Select ${isTheme ? 'a theme' : 'an extension'} to upload`,
+        pageSize: 10,
+        choices
+      })).extensionId
+
+      const extension = extensionPropeties.find(({ id }) => id === selectedId)
+      extensionId = extension.id
+      extensionDir = extension.dir
+    }
+
+    const extensionPath = path.resolve(extensionsDirectory, extensionDir)
+    logger.debug('Extension path:', extensionPath)
+
+    let extensionConfig
+    try {
+      extensionConfig = await fsEx.readJSON(path.resolve(extensionPath, 'extension-config.json'))
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        err.message = `There is no extension-config.json in ${extensionDir}`
+      }
+      throw err
+    }
+
+    if (!extensionConfig || !extensionConfig.version) {
+      throw new Error('Invalid extension-config.json')
+    }
+
+    extensionVersion = extensionConfig.version
+    logger.debug(`${isTheme ? 'Theme' : 'Extension'} version:`, extensionVersion)
+
+    try {
+      logger.debug(`Checking if ${isTheme ? 'theme' : 'extension'} ${extensionId} exists`)
+      await this.dcHttpClient.getExtension(extensionId, extensionVersion)
+    } catch (err) {
+      const message = `Error getting ${extensionId}: ${err.message || 'Unknown error'}`
+      throw new Error(message)
+    }
+
+    try {
```

Comment:
The logic of checking of the status of one version should be in the Developer Connector.
Also the patching of a version or the creating of an extension should be handeled there.

The benefit are:
- have less APIs/contact points that we need to manage 
- we have more code running on a known platform (less code on the developers computer)
- more control and faster updated on changes (because we just need to update the DC)

Reply:
The logic of version checking/creation/patching was moved to DC